### ISSUE TITLE
Java version does not support new versioning schema introduced in Java 9

### ISSUE THREAD

**Issue by dpolivaev** [knopflerfish/knopflerfish.org#8]
It should be fixed in org.osgi.framework.Version.Version(String) 
Currently Knopflerfish fails to run bundles containing Bundle-RequiredExecutionEnvironment constraints in their manifests under Java 9

**Comment by chlarsson** [comment]
Thanks for reporting this.

In Java 9 the "java.specification.version" property  has been changed so that the former major version "1" as in "1.9" has been dropped and former minor version has become major version, e.g. "9"

Knopflerfish makes checks to verify the major version is "1"

You can go around the problem for now by setting:
Forg.osgi.framework.executionenvironment="JavaSE-9"

**Comment by dpolivaev** [comment]
Unfortunately I can not use this workaround because the Version constructor throws an exception given string "9.-1" which is currently returned as java version property value.

**Comment by chlarsson** [comment]
Which OS and version of Java9 are you using?

**Comment by dpolivaev** [comment]
Windows 10
C:\Program Files\Java\jdk-9\bin>java -version
java version "9-ea"
Java(TM) SE Runtime Environment (build 9-ea+143)
Java HotSpot(TM) 64-Bit Server VM (build 9-ea+143, mixed mode)

**Comment by chlarsson** [comment]
Strange, I don't understand where the "9.-1" string comes from.

What happens if you start like this:
java -jar framework.jar -Forg.osgi.framework.executionenvironment="JavaSE-9" -init

**Comment by dpolivaev** [comment]
I have investigated what happens. It turned out that problems are in class FWProps.

System.getProperty("java.specification.version") returns "9" . 
System.getProperty("java.version") would return "9-ea", but it is not relevant.
Therefore blocks

```
          if (startPos < endPos) {
            javaVersionMinor = Integer.parseInt(javaVersion.substring(startPos, endPos));
            startPos = endPos + 1;
            endPos = startPos;
            while (endPos < max && Character.isDigit(javaVersion.charAt(endPos))) {
              endPos++;
            }
            if (startPos < endPos) {
              javaVersionMicro = Integer.parseInt(javaVersion.substring(startPos, endPos));
            }
          }

```

are ignored,
javaVersionMajor is set to 9, javaVersionMinor and javaVersionMicro remain equal to -1.
and code 

```
    setPropertyDefault(SYSTEM_PACKAGES_VERSION_PROP, Integer.toString(javaVersionMajor) + "."
        + javaVersionMinor);
```

sets this property to "9.-1" . It causes the exception in constructor of Version which I mentioned before.

Also in block 

```
    if (!props.containsKey(FRAMEWORK_EXECUTIONENVIRONMENT)) {
      final StringBuffer ee = new StringBuffer();
      // Always allow ee minimum
      ee.append("OSGi/Minimum-1.0");
      ee.append(",OSGi/Minimum-1.1");
      ee.append(",OSGi/Minimum-1.2");
      // Set up the default ExecutionEnvironment
      if (1 == javaVersionMajor) {
        for (int i = javaVersionMinor; i > 1; i--) {
          ee.append((i > 5) ? ",JavaSE-1." : ",J2SE-1.");
          ee.append(i);
        }
      }
      props.put(FRAMEWORK_EXECUTIONENVIRONMENT, ee.toString());
    }
```

for-loop is not executed and in method SystemBundle.initSystemBundle() provideCapabilityString becomes 'osgi.ee;osgi.ee=OSGi/Minimum;version:List<Version>="1.0.0,1.1.0,1.2.0"'

As a workaround I can set properties

```
-Forg.knopflerfish.framework.system.packages.version=1.7.0
-Forg.osgi.framework.system.capabilities=osgi.ee;osgi.ee=JavaSE;version:List<Version>="1.2.0,1.3.0,1.4.0,1.5.0,1.6.0,1.7.0",osgi.ee;osgi.ee=OSGi/Minimum;version:List<Version>="1.0.0,1.1.0,1.2.0"
```

because I actually need JavaSE 7 only .
Well, after I force these property values it turns out that Java9 is still full of bugs, but is another issue.

**Comment by janste63** [comment]
Fixed  in framework 8.0.2. Available in https://bintray.com/knopflerfish/master/download_file?file_path=distrib_6.0.0.snapshot-1701031635-master%2Fknopflerfish_osgi_sdk_6.0.0.snapshot-1701031635-master.jar

**Comment by dpolivaev** [comment]
How can the fixed version be used as maven / gradle dependency?

**Comment by chlarsson** [comment]
Sorry for the late reply.

The fix is in the KF 6.1 release and the framework and bundles are included in the KF 6.1 maven release repository.

All the maven artifacts are listed here
http://www.knopflerfish.org/releases/6.1.0/maven2/org/knopflerfish/kf6/KF-6.1.0_dependencyManagement.xml

The maven2 repo is browsable at:
http://www.knopflerfish.org/releases/6.1.0/maven2/org/knopflerfish/kf6/

The KF 6.1 bundles can

**Comment by dpolivaev** [comment]
Thank you. 
We still have time before java 9 release. 
And with this version it works fine.
